Readme.md
# [Zoom/Pan Options](https://foundryvtt.com/packages/zoom-pan-options/)

![GitHub release (latest SemVer)](https://img.shields.io/github/v/release/PsychoH13/ZoomPanOptions?style=for-the-badge)
![GitHub Releases](https://img.shields.io/github/downloads/PsychoH13/ZoomPanOptions/latest/total?style=for-the-badge)
![GitHub All Releases](https://img.shields.io/github/downloads/PsychoH13/ZoomPanOptions/total?style=for-the-badge&label=Downloads+total)
![Latest Supported Foundry Version](https://img.shields.io/endpoint?url=https://foundryshields.com/version?url=https://github.com/PsychoH13/ZoomPanOptions/raw/master/module.json)

FoundryVTT module to change zooming and panning through the mouse or a touchpad. It has several settings, which can be toggled individually, and stored locally (per client).

To install, browse for it in the module browser, or [directly copy the manifest link for the latest release](https://github.com/PsychoH13/ZoomPanOptions/releases/latest/download/module.json).

# Features

## Zoom around cursor
- When zooming in and out, the camera will stay focused on the cursor, like in many other applications.
- This does not affect PageUp, PageDown, or Numpad +/-. Those will still zoom focused on the center of the screen.

## Middle-mouse to pan
- Holding the middle mouse button and dragging around will pan around the map, just like the right mouse button, instead of showing the "auto scroll" icon (which is useless in Foundry).
- This will mimic the default right-mouse panning, though it will ignore tokens, tiles, etc (instead of selecting them).

## Disable zoom rounding
- Disables default Foundry behavior, which rounds zoom to the nearest 1%. Will make zooming smoother, especially for touchpad users.

## Minimum/Maximum Zoom Override
- Override for the minimum and maximum zoom scale limits. 3 is the Foundry default - you can't zoom in to get a bigger than a x3 scale closeup, or zoom out to get a smaller than a x0.3 "wide shot" of the scene. For example, if you change this to 6, you'll be able to zoom in twice as close and zoom out twice as far.  A value of 10 is usually enough to fill the screen with a single token or with the entire scene

## Pan/Zoom Scroll Wheel Mode:
You can choose any combination of Shift, Ctrl, Alt, and Meta, as well as no modifiers together with the scroll wheel to perform the following actions:
- Zoom
- Pan
- Pan Horizontally (useful if your mouse doesn't support horizontal scrolling)
- Rotate element
- Snap Rotate element

You can also disable any of these actions.

## Pan speed multiplier
- Only used in touchpad and alternative modes. Multiplies pan speed. Defaults to 1, which should be close to the pan speed when right-click-dragging the canvas.

## Zoom speed multiplier
- Useful if your zoom is too sensitive, or not sensitive enough.
- Set to 0 for default Foundry behavior (5% zoom per mouse tick, always).
- Set to 1 for zooming based on scroll delta, which should be similar to default zoom for most common mouse types, but will feel smoother for some touchpads.
- Set to 0.1 for slower zooming, or 10 for faster zooming.
- Technically you can give this a negative value to flip your zoom directions, if you're an oddball.

## Invert vertical scroll
- Only used in touchpad and alternative modes. If set to true, you will scroll up when dragging/scrolling down.

# Credits

Thanks to mrkwnzl#7407 for the touchpad support and testing!

Thanks to akrigline/Calego for merging with CanvasScroll and helping with libwrapper and workflows!

Thanks to TPNils for the touchpad auto-detection help!

# [Cursor Zoom](https://github.com/itamarcu/CursorZoom)
My old foundry module, that only had the "Zoom around cursor" feature, and did not allow configuring it in the settings for each player.
If you were a user of Cursor Zoom, please uninstall it and install this module. Your players will be able to each set their own preference.


Changelog.md
# Changelog
All notable changes to this project will be documented in this file.

The format is loosely based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.8.0] - 2022-07-09
- Replace modes with setting individual key modifier combinations per scroll setting.

## [1.7.2] - 2022-07-02
- FoundryVTT V10 compatibility
- "Disable Zoom Rounding" had to be changed;  please inform me if it's worse now.

## [1.7.0] - 2022-02-25
- Split "auto-detect touchpad" setting out of pan/zoom mode, and made it false by default
- Added settings to change pad and shift values when dragging at the edge of the screen (#33)

## [1.6.10] - 2022-02-08
- Updated libWrapper shim code

## [1.6.9] - 2021-12-26
- Explicitly marked libWrapper as a dependency, and updated libWrapper shim code

## [1.6.8] - 2021-12-24
- Fixed compatibility with Foundry v9.232 (#35, #37) (thank you @caewok for the contribution!)
- Marked as compatible with Foundry v9 stable

## [1.6.5] - 2021-07-06
- Fixed minimum/maximum zoom not carrying over after refreshing ([#28](https://github.com/itamarcu/ZoomPanOptions/issues/28))
- Updated default min/max zoom scale to be just like the Foundry default - i.e. `3.0`.
- Fixed rotation not working on Firefox ([#29](https://github.com/itamarcu/ZoomPanOptions/issues/29))

## [1.6.3] - 2021-07-06
- Added automatic detection of touchpad movements in "Default" mode
- Added "DefaultMouse" mode which prevents this automatic detection

## [1.6.1] - 2021-05-29
- Added middle mouse pan setting
- Added changelog and license to repository

## [1.5.3] - 2021-02-05
- Added invert vertical scroll setting
- Introduced new bugs and promptly fixed them

## [1.4.0] - 2020-12-05
- Added min/max zoom override setting
- Fixed compatibility with Foundry v0.7.x

## [1.3.7] - 2020-09-17
- Added ability to use alt/option/meta key
- Fixed some minor bugs
- Improved readme
- Improved file structure

## [1.3.0] - 2020-09-16
- Integrated Canvas Scroll (thanks @akrigline!)
- Reworked to use 3 modes
- Added LibWrapper as optional dependency

## 1.0.0 - 2020-09-15
- Created the module, with initial feature set

## See also: [Unreleased]

[1.3.0]: https://github.com/itamarcu/ZoomPanOptions/compare/1.0.0...1.3.0
[1.3.7]: https://github.com/itamarcu/ZoomPanOptions/compare/1.3.0...1.3.7
[1.4.0]: https://github.com/itamarcu/ZoomPanOptions/compare/1.3.7...1.4.0
[1.5.3]: https://github.com/itamarcu/ZoomPanOptions/compare/1.4.0...1.5.3
[1.6.1]: https://github.com/itamarcu/ZoomPanOptions/compare/1.5.3...1.6.1
[1.6.3]: https://github.com/itamarcu/ZoomPanOptions/compare/1.6.1...1.6.3
[1.6.5]: https://github.com/itamarcu/ZoomPanOptions/compare/1.6.3...1.6.5
[1.6.8]: https://github.com/itamarcu/ZoomPanOptions/compare/1.6.5...1.6.8
[1.6.9]: https://github.com/itamarcu/ZoomPanOptions/compare/1.6.8...1.6.9
[1.6.10]: https://github.com/itamarcu/ZoomPanOptions/compare/1.6.9...1.6.10
[1.7.0]: https://github.com/itamarcu/ZoomPanOptions/compare/1.6.10...1.7.0
[1.7.2]: https://github.com/itamarcu/ZoomPanOptions/compare/1.7.2...1.7.2
[Unreleased]: https://github.com/itamarcu/ZoomPanOptions/compare/1.7.2...HEAD


scripts/zoom-pan-options.js
import { libWrapper } from './libwrapper-shim.js'

const MODULE_ID = 'zoom-pan-options'

// note: 'Default' is the old name for 'AutoDetect'
// 'DefaultMouse' is the old name for 'Mouse'

function getSetting (settingName) {
  return game.settings.get(MODULE_ID, settingName)
}

function checkRotationRateLimit (layer) {
  const hasTarget = layer.options?.controllableObjects ? layer.controlled.length : !!layer._hover
  if (!hasTarget)
    return false
  const t = Date.now()
  const rate_limit = isNewerVersion(game.version, '9.231')
    ? game.mouse.MOUSE_WHEEL_RATE_LIMIT
    : game.keyboard.constructor.MOUSE_WHEEL_RATE_LIMIT

  if ((t - game.keyboard._wheelTime) < rate_limit)
    return false
  game.keyboard._wheelTime = t
  return true
}

function getSettingsByModifiers() {
  let result = {}

  const zoomModifiers = getSetting('zoom-modifier')
  result[zoomModifiers] = 'zoom'

  const rotateModifiers = getSetting('rotate-modifier')
  if (!result[rotateModifiers])
    result[rotateModifiers] = 'rotate'

  const rotateSnapModifiers = getSetting('rotate-snap-modifier')
  if (!result[rotateSnapModifiers])
    result[rotateSnapModifiers] = 'rotate-snap'

  const panModifiers = getSetting('pan-modifier')
  if (!result[panModifiers])
    result[panModifiers] = 'pan'

  const horizontalPanModifiers = getSetting('horizontal-pan-modifier')
  if (!result[horizontalPanModifiers])
    result[horizontalPanModifiers] = 'horizontal-pan'

  return result
}

function modifiersForEvent(event) {
  let array = []
  if (event.shiftKey)
    array.push('shift')

  if (event.ctrlKey)
    array.push('ctrl')

  if (event.altKey)
    array.push('alt')

  if (event.metaKey)
    array.push('meta')

  return array.join('-')
}

/**
 * (note: return value is meaningless here)
 */
function _onWheel_Override (event) {
  const ctrlOrMeta = event.ctrlKey || event.metaKey  // meta key (cmd on mac, winkey in windows) will behave like ctrl

  const settingsByModifiers = getSettingsByModifiers()
  const eventModifiers = modifiersForEvent(event)
  const scrollMode = settingsByModifiers[eventModifiers]

  // Prevent zooming the entire browser window
  if (ctrlOrMeta) {
    event.preventDefault()
  }

  // Take no actions if the canvas is not hovered
  if (!canvas?.ready)
    return

  const hover = document.elementFromPoint(event.clientX, event.clientY)
  if (!hover || (hover.id !== 'board'))
    return

  event.preventDefault()

  const layer = canvas.activeLayer

  // Case 1 - rotate stuff
  if (layer instanceof PlaceablesLayer) {
    const deltaY = event.wheelDelta !== undefined ? event.wheelDelta
      // wheelDelta is undefined in firefox
      : event.deltaY

    if (scrollMode == 'rotate' || scrollMode == 'rotate-snap') {
      return checkRotationRateLimit(layer) && layer._onMouseWheel({
        deltaY: deltaY,
        shiftKey: scrollMode == 'rotate-snap',
      })
    }
  }

  // Case 2 - zoom the canvas
  if (scrollMode == 'zoom')
    return zoom(event)

  // Cast 3 - pan the canvas horizontally (shift+scroll)
  if (scrollMode == 'horizontal-pan') {
    // noinspection JSSuspiciousNameCombination
    return panWithMultiplier({
      deltaX: event.deltaY,
    })
  }

  // Case 4 - pan the canvas in the direction of the mouse/touchpad event
  if (scrollMode == 'pan')
    return panWithMultiplier(event)
}

/**
 * note - this is useless once we get to V10
 */
function _constrainView_Override ({ x, y, scale } = {}) {
  const d = canvas.dimensions

  // Constrain the maximum zoom level
  if (Number.isNumeric(scale) && scale !== canvas.stage.scale.x) {
    const max = CONFIG.Canvas.maxZoom
    const ratio = Math.max(d.width / window.innerWidth, d.height / window.innerHeight, max)
    // override changes are just for this part:
    if (getSetting('disable-zoom-rounding')) scale = Math.clamped(scale, 1 / ratio, max)
    else scale = Math.round(Math.clamped(scale, 1 / ratio, max) * 100) / 100
  } else {
    scale = canvas.stage.scale.x
  }

  // Constrain the pivot point using the new scale
  if (Number.isNumeric(x) && x !== canvas.stage.pivot.x) {
    const padw = 0.4 * (window.innerWidth / scale)
    x = Math.clamped(x, -padw, d.width + padw)
  } else x = canvas.stage.pivot.x
  if (Number.isNumeric(y) && x !== canvas.stage.pivot.y) {
    const padh = 0.4 * (window.innerHeight / scale)
    y = Math.clamped(y, -padh, d.height + padh)
  } else y = canvas.stage.pivot.y

  // Return the constrained view dimensions
  return { x, y, scale }
}

/**
 * Will zoom around cursor, and based on delta.
 */
function zoom (event) {
  const multiplier = getSetting('zoom-speed-multiplier')
  let dz = -event.deltaY * 0.0005 * multiplier + 1
  // default foundry behavior if zoom-speed-multiplier is 0
  if (multiplier === 0) dz = event.deltaY < 0 ? 1.05 : 0.95

  if (!getSetting('zoom-around-cursor')) {
    canvas.pan({ scale: dz * canvas.stage.scale.x })
    return
  }

  const scale = dz * canvas.stage.scale.x
  const d = canvas.dimensions
  const max = CONFIG.Canvas.maxZoom
  const min = 1 / Math.max(d.width / window.innerWidth, d.height / window.innerHeight, max)

  if (scale > max || scale < min) {
    canvas.pan({ scale: scale > max ? max : min })
    console.log('Zoom/Pan Options |', `scale limit reached (${scale}).`)
    return
  }

  // Acquire the cursor position transformed to Canvas coordinates
  const t = canvas.stage.worldTransform
  const dx = ((-t.tx + event.clientX) / canvas.stage.scale.x - canvas.stage.pivot.x) * (dz - 1)
  const dy = ((-t.ty + event.clientY) / canvas.stage.scale.y - canvas.stage.pivot.y) * (dz - 1)
  const x = canvas.stage.pivot.x + dx
  const y = canvas.stage.pivot.y + dy
  canvas.pan({ x, y, scale })
}

function panWithMultiplier (event) {
  const multiplier = (1 / canvas.stage.scale.x) * getSetting('pan-speed-multiplier')
  const invertVerticalScroll = getSetting('invert-vertical-scroll') ? -1 : 1
  const x = canvas.stage.pivot.x + event.deltaX * multiplier
  const y = canvas.stage.pivot.y + event.deltaY * multiplier * invertVerticalScroll
  canvas.pan({ x, y })
}

function disableMiddleMouseScrollIfMiddleMousePanIsActive (isActive) {
  if (isActive) {
    // this will prevent middle-click from showing the scroll icon
    document.body.onmousedown__disabled = document.body.onmousedown
    document.body.onmousedown = function (e) { if (e.button === 1) return false }
  } else {
    document.body.onmousedown = document.body.onmousedown__disabled
  }
}

function _handleMouseDown_Wrapper (wrapped, ...args) {
  if (!getSetting('middle-mouse-pan')) return wrapped(...args)
  const event = args[0]
  if (event.data.originalEvent.button === 0) return wrapped(...args) // left-click
  if (event.data.originalEvent.button !== 1) return // additional buttons other than middle click - still ignoring!

  // Middle-mouse click will *only* pan;  ignoring anything else on the canvas
  const mim = canvas.mouseInteractionManager
  if (![mim.states.HOVER, mim.states.CLICKED, mim.states.DRAG].includes(mim.state)) return wrapped(...args)
  canvas.currentMouseManager = mim

  // Update event data
  event.data.object = mim.object
  event.data.origin = event.data.getLocalPosition(mim.layer)

  // piggy-backing off of the right-mouse-drag code, for lack of a better option
  const action = 'clickRight'
  if (!mim.can(action, event)) return
  event.stopPropagation()
  mim._dragRight = true

  // Upgrade hover to clicked
  if (mim.state === mim.states.HOVER) mim.state = mim.states.CLICKED
  if (CONFIG.debug.mouseInteraction) console.log(`${mim.object.constructor.name} | ${action}`)

  // Trigger callback function
  mim.callback(action, event)

  // Activate drag handlers
  if ((mim.state < mim.states.DRAG) && mim.can('dragRight', event)) {
    mim._activateDragEvents()
  }
}

/**
 * Changes from original function:
 * `pad` value and `shift` divisor are both customizable instead of being the default of 25 and 2.
 */
function _onDragCanvasPan_override (event) {
  // Throttle panning by 200ms
  const now = Date.now()
  if (now - (this._panTime || 0) <= 200) return
  this._panTime = now

  // Shift by a few grid spaces at a time
  const { x, y } = event
  const pad = getSetting('pad-value-when-dragging')
  const shift = (this.dimensions.size * getSetting('shift-value-when-dragging')) / this.stage.scale.x

  // Shift horizontally
  let dx = 0
  if (x < pad) dx = -shift
  else if (x > window.innerWidth - pad) dx = shift

  // Shift vertically
  let dy = 0
  if (y < pad) dy = -shift
  else if (y > window.innerHeight - pad) dy = shift

  // Enact panning
  if (dx || dy) return this.animatePan({ x: this.stage.pivot.x + dx, y: this.stage.pivot.y + dy, duration: 200 })
}

const migrateOldSettings = () => {
  const mode = getSetting('pan-zoom-mode')
  if (mode === 'DefaultMouse') {
    console.log(`Zoom/Pan Options | Migrating old setting 'pan-zoom-mode': 'DefaultMouse' to 'Mouse'`)
    game.settings.set('zoom-pan-options', 'pan-zoom-mode', 'Mouse')
  }
  if (mode === 'Default') {
    console.log(`Zoom/Pan Options | Migrating old setting 'pan-zoom-mode': 'Default' to 'Mouse'`)
    game.settings.set('zoom-pan-options', 'pan-zoom-mode', 'Mouse')
  }

  if (mode === 'Mouse') {
    console.log(`Zoom/Pan Options | Migrating old setting 'zoom-pan-options': to zoom-modifier: '', rotate-modifier: 'ctrl', rotate-snap-modifier: 'shift', horizontal-pan-modifier: 'disabled', pan-modifier: 'disabled'`)
    game.settings.set('zoom-modifier', '')
    game.settings.set('rotate-modifier', 'ctrl')
    game.settings.set('rotate-snap-modifier', 'shift')
    game.settings.set('horizontal-pan-modifier', 'disabled')
    game.settings.set('pan-modifier', 'disabled')
  } else if (mode === 'Touchpad') {
    console.log(`Zoom/Pan Options | Migrating old setting 'zoom-pan-options': to zoom-modifier: 'ctrl', rotate-modifier: 'shift', rotate-snap-modifier: 'shift-ctrl', horizontal-pan-modifier: 'disabled', pan-modifier: 'disabled'`)
    game.settings.set('zoom-modifier', 'ctrl')
    game.settings.set('rotate-modifier', 'shift')
    game.settings.set('rotate-snap-modifier', 'shift-ctrl')
    game.settings.set('horizontal-pan-modifier', 'disabled')
    game.settings.set('pan-modifier', 'disabled')
  } else if (mode === 'Alternative') {
    console.log(`Zoom/Pan Options | Migrating old setting 'zoom-pan-options': to zoom-modifier: 'ctrl', rotate-modifier: 'shift-alt', rotate-snap-modifier: 'ctrl-alt', horizontal-pan-modifier: 'shift', pan-modifier: ''`)
    game.settings.set('zoom-modifier', 'ctrl')
    game.settings.set('rotate-modifier', 'shift-alt')
    game.settings.set('rotate-snap-modifier', 'ctrl-alt')
    game.settings.set('horizontal-pan-modifier', 'shift')
    game.settings.set('pan-modifier', '')
  }
}

Hooks.on('init', function () {
  console.log('Initializing Zoom/Pan Options')
  game.settings.register(MODULE_ID, 'zoom-around-cursor', {
    name: 'Zoom around cursor',
    hint: 'Center zooming around cursor. Does not apply to zooming with pageup or pagedown.',
    scope: 'client',
    config: true,
    default: true,
    type: Boolean,
  })
  game.settings.register(MODULE_ID, 'middle-mouse-pan', {
    name: 'Middle-mouse to pan',
    hint: 'Middle mouse press will pan the canvas, instead of the default of doing nothing.',
    scope: 'client',
    config: true,
    default: false,
    type: Boolean,
    onChange: disableMiddleMouseScrollIfMiddleMousePanIsActive,
  })
  game.settings.register(MODULE_ID, 'disable-zoom-rounding', {
    name: 'Disable zoom rounding',
    hint:
      'Disables default Foundry behavior, which rounds zoom to the nearest 1%. Will make zooming smoother, especially for touchpad users.',
    scope: 'client',
    config: true,
    default: true,
    type: Boolean,
  })
  game.settings.register(MODULE_ID, 'min-max-zoom-override', {
    name: 'Minimum/maximum zoom override',
    hint: 'Override for the minimum and maximum zoom scale limits. 3 is the Foundry default (x3 and x0.333 scaling).',
    scope: 'client',
    config: true,
    default: CONFIG.Canvas.maxZoom, // 3.0
    type: Number,
    onChange: value => {
      CONFIG.Canvas.maxZoom = value
    },
  })

  const modifierChoices = {
    '': '-',
    'shift': 'Shift',
    'ctrl': 'Ctrl',
    'alt': 'Alt',
    'meta': 'Meta',
    'shift-ctrl': 'Shift + Ctrl',
    'shift-alt': 'Shift + Alt',
    'shift-meta': 'Shift + Meta',
    'ctrl-alt': 'Ctrl + Alt',
    'ctrl-meta': 'Ctrl + Meta',
    'alt-meta': 'Alt + Meta',
    'shift-ctrl-alt': 'Shift + Ctrl + Alt',
    'shift-ctrl-meta': 'Shift + Ctrl + Meta',
    'shift-alt-meta': 'Shift + Alt + Meta',
    'ctrl-alt-meta': 'Ctrl + Alt + Meta',
    'shift-ctrl-alt-meta': 'Shift + Ctrl + Alt + Meta',
    'disabled': 'Disabled',
  }

  game.settings.register(MODULE_ID, 'zoom-modifier', {
    name: 'Scroll to Zoom Modifiers',
    hint: 'Key to hold to zoom.',
    scope: 'client',
    config: true,
    type: String,
    choices: modifierChoices,
    default: '',
  })
  game.settings.register(MODULE_ID, 'pan-modifier', {
    name: 'Scroll to Pan Modifiers',
    hint: 'Key to hold to pan.',
    scope: 'client',
    config: true,
    type: String,
    choices: modifierChoices,
    default: 'disabled',
  })
  game.settings.register(MODULE_ID, 'horizontal-pan-modifier', {
    name: 'Scroll to Pan Horizontally Modifiers',
    hint: 'Key to hold to pan horizontally. Useful if your mouse doesn\'t support horizontal scrolling.',
    scope: 'client',
    config: true,
    type: String,
    choices: modifierChoices,
    default: 'disabled',
  })
  game.settings.register(MODULE_ID, 'rotate-modifier', {
    name: 'Scroll to Rotate Modifiers',
    hint: 'Key to hold to rotate an item.',
    scope: 'client',
    config: true,
    type: String,
    choices: modifierChoices,
    default: 'ctrl',
  })
  game.settings.register(MODULE_ID, 'rotate-snap-modifier', {
    name: 'Scroll to Snap Rotate Modifiers',
    hint: 'Key to hold to rotate an item and snap to the grid.',
    scope: 'client',
    config: true,
    type: String,
    choices: modifierChoices,
    default: 'shift',
  })

  game.settings.register(MODULE_ID, 'zoom-speed-multiplier', {
    name: 'Zoom speed',
    hint:
      'Multiplies zoom speed, affecting scaling speed. 0.1 or 10 might be better for some touchpads. 0 for default Foundry behavior (which ignores scroll "intensity", and feels worse for touchpads).',
    scope: 'client',
    config: true,
    default: 0,
    type: Number,
  })
  game.settings.register(MODULE_ID, 'pan-speed-multiplier', {
    name: 'Pan speed',
    hint:
      'Multiplies pan speed. Defaults to 1, which should be close to the pan speed when right-click-dragging the canvas.',
    scope: 'client',
    config: true,
    default: 1,
    type: Number,
  })
  game.settings.register(MODULE_ID, 'invert-vertical-scroll', {
    name: 'Invert vertical scroll',
    hint: 'If set to true, you will scroll up when dragging/scrolling down.',
    scope: 'client',
    config: true,
    default: false,
    type: Boolean,
  })
  game.settings.register(MODULE_ID, 'pad-value-when-dragging', {
    name: '"pad" value when dragging something to the edge of the screen',
    hint:
      'When holding down the cursor and moving it towards the edge of the screen, the canvas will pan.  "pad" is the distance that will trigger it. Foundry default is 50px.',
    scope: 'client',
    config: true,
    default: 50,
    type: Number,
  })
  game.settings.register(MODULE_ID, 'shift-value-when-dragging', {
    name: '"shift" value when dragging something to the edge of the screen',
    hint:
      'When holding down the cursor and moving it towards the edge of the screen, the canvas will pan.  "shift" is the panning distance in tiles. Foundry default is 3 tiles.',
    scope: 'client',
    config: true,
    default: 3,
    type: Number,
  })
  migrateOldSettings()
})

Hooks.once('setup', function () {
  const wheelPrototype = isNewerVersion(game.version, '9.231')
    ? 'MouseManager.prototype._onWheel'
    : 'KeyboardManager.prototype._onWheel'

  libWrapper.register(
    MODULE_ID,
    wheelPrototype,
    (event) => {
      return _onWheel_Override(event)
    },
    'OVERRIDE',
  )
  if (isNewerVersion(game.version, '10.269')) {
    // temporary solution, due to constrainView becoming private in V10.
    // see https://github.com/foundryvtt/foundryvtt/issues/7382
    // or see https://github.com/foundryvtt/foundryvtt/issues/7189
    Hooks.on('canvasPan', (board, constrained) => {
      if (!getSetting('disable-zoom-rounding')) return

      const d = canvas.dimensions
      const max = CONFIG.Canvas.maxZoom
      const ratio = Math.max(d.width / window.innerWidth, d.height / window.innerHeight, max)
      constrained.scale = Math.clamped(constrained.scale, 1 / ratio, max)

      board.stage.scale.set(constrained.scale, constrained.scale)
      board.updateBlur(constrained.scale)
    })
  } else {
    libWrapper.register(
      MODULE_ID,
      'Canvas.prototype._constrainView',
      (obj) => {
        return _constrainView_Override(obj)
      },
      'OVERRIDE', // only overrides a tiny part of the function... would be nice if foundry made it more modular
    )
  }
  libWrapper.register(
    MODULE_ID,
    'Canvas.prototype._onDragCanvasPan',
    _onDragCanvasPan_override,
    'OVERRIDE', // (same as above)
  )
  libWrapper.register(
    MODULE_ID,
    'MouseInteractionManager.prototype._handleMouseDown',
    function (wrapped, ...args) {
      return _handleMouseDown_Wrapper.bind(this)(wrapped, ...args)
    },
    'MIXED', // only overrides if it's a middle click
  )
  disableMiddleMouseScrollIfMiddleMousePanIsActive(getSetting('middle-mouse-pan'))
  CONFIG.Canvas.maxZoom = getSetting('min-max-zoom-override')
  console.log('Done setting up Zoom/Pan Options.')
})
